### PR TITLE
Added unit tests for suppression matching

### PR DESCRIPTION
**Pull request by duncanp-sonar**
I made some minor changes to the product code to make it more testable - I moved the suppression matching code into a separate class and added an interface for the live issue factory.

### CODE REVIEW

File: src/Integration.Vsix/Suppression/SuppressionHandler.cs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            if (serverIssuesProvider == null)
+            {
+                throw new ArgumentNullException(nameof(serverIssuesProvider));
+            }
+
+            this.liveIssueFactory = liveIssueFactory;
+            this.serverIssuesProvider = serverIssuesProvider;
+        }
+
+        public bool ShouldIssueBeReported(SyntaxTree syntaxTree, Diagnostic diagnostic)
```
Comment (duncanp-sonar) [inline src/Integration.Vsix/Suppression/SuppressionHandler.cs]: The logic in this method hasn't changed - it's just been moved from the _SuppressionManager_ class.

File: src/Integration.Vsix/Suppression/SuppressionManager.cs [hunk 2]
```diff
@@ -96,43 +96,14 @@ private void OnSolutionBindingChanged(object sender, ActiveSolutionBindingEventA
         private bool ShouldIssueBeReported(SyntaxTree syntaxTree, Diagnostic diagnostic)
         {
             // This method is called for every analyzer issue that is raised so it should be fast.
-            if (!diagnostic.Location.IsInSource &&
-                diagnostic.Location != Location.None)
-            {
-                return true;
-            }
 
             if (activeSolutionBoundTracker == null ||
```
Comment (duncanp-sonar) [inline src/Integration.Vsix/Suppression/SuppressionManager.cs]: I left this check in the _SuppressionManager_ so that the suppression handler doesn't need to know about the solution tracker.

File: src/Integration.Vsix.UnitTests/Integration.Vsix.UnitTests.csproj [hunk 3]
```diff
@@ -323,6 +323,10 @@
       <Project>{03278c1a-ee78-4fa7-a5b0-1ab0a81ea76f}</Project>
       <Name>Integration</Name>
     </ProjectReference>
+    <ProjectReference Include="..\SonarQube.Client\SonarQube.Client.csproj">
```
Comment (Evangelink) [inline src/Integration.Vsix.UnitTests/Integration.Vsix.UnitTests.csproj]: You probably need the same change for VS2017.

File: src/Integration.Vsix/Suppression/LiveIssue.cs [hunk 4]
```diff
@@ -24,10 +24,10 @@
 namespace SonarLint.VisualStudio.Integration.Vsix.Suppression
 {
     /// <summary>
-    ///     Information about a single Roslyn issue, decorated with the extra information required to map it to a
-    ///     SonarQube issue.
+    /// Information about a single Roslyn issue, decorated with the extra information required to map it to a
+    /// SonarQube issue.
     /// </summary>
-    internal class LiveIssue
+    public class LiveIssue
```
Comment (Evangelink) [inline src/Integration.Vsix/Suppression/LiveIssue.cs]: Why this change?

File: src/Integration.Vsix.UnitTests/Integration.Vsix.UnitTests.csproj (hunk 3 above)
Comment (duncanp-sonar) [reply src/Integration.Vsix.UnitTests/Integration.Vsix.UnitTests.csproj]: +1

File: src/Integration.Vsix/Suppression/LiveIssue.cs (hunk 4 above)
Comment (duncanp-sonar) [reply src/Integration.Vsix/Suppression/LiveIssue.cs]: ILiveIssueFactory needs to be public so Moq can mock it so LiveIssue needs to be public too as it's returned by the interface.

File: src/Integration.Vsix/Suppression/SuppressionHandler.cs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            // 2. Same component, same file, same error code, same line             // tolerant to code on the line changing e.g. var rename
+
+            // As a minimum, the project, file and rule id must match
+            var issuesInFile = serverIssuesProvider.GetSuppressedIssues(liveIssue.ProjectGuid, liveIssue.IssueFilePath);
+            if (issuesInFile == null)
+            {
+                return true;
+            }
+
+            // TODO: rule repository?
```
Comment (SonarTech) [inline src/Integration.Vsix/Suppression/SuppressionHandler.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by duncanp-sonar** [comment]
Build failure: the unrelated test _Exceptions_From_SonarQubeService_Are_Written_To_Output_ was failing for some reason. I've rebased and the test is now passing.
